New error for Output:Table:Monthly and Annual don't have proper order for aggregation types

JasonGlazer opened the pull request:
Add new error message for Output:Table:Monthly and Output:Table:Annual when the advanced aggregation types (such as MaximumDuringHoursShown and SumOrAverageDuringHoursShown) are used when the corresponding aggregation type is not used in an earlier column (such as Maximum or HoursNonZero). 

This resolves #5911

 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

JasonGlazer committed:
Initial checkin

JasonGlazer committed:
Merge branch 'develop' into NewErrorWhenHoursShownInFirstColumnOfTable

JasonGlazer committed:
Add warning message for Annual reports with invalid ordering of aggregation types.

JasonGlazer committed:
Add message for Monthly tables that have invalid aggregation type ordering.

JasonGlazer committed:
Add new unit test.

JasonGlazer committed:
Changed from switch to if block to remove warning from gcc compiler.

JasonGlazer commented:
I added a defect file to show the new warning messages. See [here](https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/5000s/5911).

Myoldmopar commented on tst/EnergyPlus/unit/OutputReportTabularAnnual.unit.cc:
```diff
@@ -272,3 +272,40 @@ TEST_F(EnergyPlusFixture, OutputReportTabularAnnual_columnHeadersToTitleCase)
+		"Misc Facility Electric Energy, !- Variable or Meter 3 Name",
+		"SumOrAverage, !- Aggregation Type for Variable or Meter 3",
+		"0; !- field Digits After Decimal 3",
+	} );
```
Do you think it's worth adding a test for `Output:Table:Monthly`?  It's hard to tell from the diff exactly what functional changes were made to the code, so the more unit tests the better.

JasonGlazer replied:
Sure, that is a good idea.

Myoldmopar commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -6395,244 +6399,272 @@ namespace OutputReportTabular {
+			}
+			if ( missingHourAggError ) {
+				ShowWarningError( "The Output:Table:Monthly report named=\"" + MonthlyInput( iInput ).name + "\" has a --DuringHoursShown aggregation type for a column without a previous field that uses one of the Hour-- aggregation types. The report will not be generated." );
+			}
```
If we aren't generating the report, is it worth just shutting it down with a fatal here?  I'd hate for a user to run an hour long simulation and quietly not get the report they were expecting.  Seems severe enough here to call fatal to me, but I don't feel _that_ strongly about it.

JasonGlazer replied:
I don't feel strongly about it either. It seems like most of the messaging related to the monthly reporting were warnings but I can make this one a fatal if you want.

JasonGlazer replied:
Looking at the code, the testing that I added occurs during the writing of the monthly and annual tables near the end of the simulation. I probably should put them at the beginning just after getinput.

Myoldmopar reviewed (commented):
Made a few comments here.  Nothing major but I'd like to hear your thoughts on the two comments before continuing on this.

JasonGlazer committed:
Merge branch 'develop' into NewErrorWhenHoursShownInFirstColumnOfTable

JasonGlazer committed:
Moved aggregation order checking to after getinput and added unit test.

JasonGlazer commented:
@Myoldmopar I made the changes you suggested. It ended up moving around a bunch of code but the checking algorithm is the same it is just being used at a different time.

JasonGlazer committed:
Remove unused variable.

JasonGlazer committed:
Merge branch 'develop' into NewErrorWhenHoursShownInFirstColumnOfTable

JasonGlazer committed:
Merge branch 'develop' into NewErrorWhenHoursShownInFirstColumnOfTable

JasonGlazer commented:
Merged only hoping that the CI issues were a fluke since I could not reproduce them locally.

JasonGlazer committed:
Add DisplayString calls in new routines to help diagnose CI issues.

JasonGlazer committed:
Check to make sure not design day only simulation

JasonGlazer committed:
Merge branch 'develop' into NewErrorWhenHoursShownInFirstColumnOfTable

JasonGlazer committed:
and for monthy reports

JasonGlazer committed:
Remove the DisplayString calls used a diagnostic tool.

JasonGlazer commented:
Turns out the reason I could not replicate the issue was that I wasn't forcing the design day runs to occur like the CI does.

mjwitte commented:
Duh - of course!

JasonGlazer committed:
Merge branch 'develop' into NewErrorWhenHoursShownInFirstColumnOfTable

JasonGlazer committed:
Remove unusued variables

JasonGlazer commented:
@Myoldmopar this is finally ready for review again. I added a unit test and moved this so that it was at the beginning of the simulation and threw a fatal so that it would not waste the users time as you suggested.

Myoldmopar commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -695,12 +695,16 @@ namespace OutputReportTabular {
+			if ( isInvalidAggregationOrder( ) ) {
+				ShowFatalError( "OutputReportTabular: Invalid aggregations detected, no simulation performed." );
+			}
+			GetInputFuelAndPollutionFactors( );
```
I'm not going to sweat it here, but it looks like your visual studio is set up to add blank padding for function calls with no args.

JasonGlazer replied:
Thanks

Myoldmopar commented on src/EnergyPlus/OutputReportTabularAnnual.cc:
```diff
@@ -246,6 +249,65 @@ namespace EnergyPlus {
+			if ( invalidAggregationOrderFound ) {
+				ShowFatalError( "OutputReportTabularAnnual: Invalid aggregations detected, no simulation performed." );
+			}
+		}
```
If this were in a more performance-centric section of the code, I'd push on why you even need the intermediate variable and why not just call the fatal from the loop above.  But I won't now.

JasonGlazer replied:
Just so all instances of this issues are reported in the error file prior to the fatal.

JasonGlazer commented:
@Myoldmopar I assume that I should make a new pull request when I re-fix this issue? I can continue using the same branch or should I make a new one of those also?

Myoldmopar commented:
As long as you pull it up to the latest develop then it doesn't matter either way.  Same branch is fine.

JasonGlazer commented:
I tried with the existing branch but it seemed to be carrying along some baggage of other peoples changes so I just created a new branch instead. See #6101
